src/planner/partialize.c

```diff
@@ -185,6 +197,652 @@ partialize_agg_paths(RelOptInfo *rel)
+								 double d_num_groups, GroupPathExtraData *extra_data)
+{
+	Query *parse = root->parse;
+
+	/* Determine costs for aggregations */
+	AggClauseCosts *agg_partial_costs = &extra_data->agg_partial_costs;
+
+	Path *hash_path = (Path *) create_agg_path(root,
+											   path->parent,
+											   path,
+											   target,
+											   AGG_HASHED,
+											   AGGSPLIT_INITIAL_SERIAL,
+											   parse->groupClause,
+											   NIL,
+											   agg_partial_costs,
+											   d_num_groups);
+	return hash_path;
+}
+
+/*
+ * Add partially aggregated subpath
+ */
+static void
+add_partially_aggregated_subpaths(PlannerInfo *root, Path *parent_path,
+								  PathTarget *partial_grouping_target, double d_num_groups,
+								  GroupPathExtraData *extra_data, bool can_sort, bool can_hash,
+								  Path *subpath, List **sorted_paths, List **hashed_paths)
+{
+	/* Translate targetlist for partition */
+	AppendRelInfo *appinfo = ts_get_appendrelinfo(root, subpath->parent->relid, false);
+	PathTarget *chunktarget = copy_pathtarget(partial_grouping_target);
+	chunktarget->exprs =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) chunktarget->exprs, 1, &appinfo));
+
+	/* In declarative partitioning planning, this is done by appy_scanjoin_target_to_path */
+	Assert(list_length(subpath->pathtarget->exprs) == list_length(parent_path->pathtarget->exprs));
+	subpath->pathtarget->sortgrouprefs = parent_path->pathtarget->sortgrouprefs;
+
+	if (can_sort)
+	{
+		*sorted_paths = lappend(*sorted_paths,
+								create_sorted_partially_agg_path(root,
+																 subpath,
+																 chunktarget,
+																 d_num_groups,
+																 extra_data));
+	}
+
+	if (can_hash)
+	{
+		*hashed_paths = lappend(*hashed_paths,
+								create_hashed_partially_agg_path(root,
+																 subpath,
+																 chunktarget,
+																 d_num_groups,
+																 extra_data));
+	}
+}
+
+/*
+ * Generate a total aggregation path for partial aggregations
+ */
+static void
+generate_agg_pushdown_path(PlannerInfo *root, Path *cheapest_total_path, RelOptInfo *output_rel,
+						   RelOptInfo *partially_grouped_rel, PathTarget *grouping_target,
+						   PathTarget *partial_grouping_target, bool can_sort, bool can_hash,
+						   double d_num_groups, GroupPathExtraData *extra_data)
+{
+	/* Get subpaths */
+	List *subpaths = get_subpaths_from_append_path(cheapest_total_path, false);
+
+	/* No subpaths available or unsupported append node */
+	if (subpaths == NIL)
+		return;
+
+	/* Generate agg paths on top of the append children */
+	List *sorted_subpaths = NIL;
+	List *hashed_subpaths = NIL;
+
+	ListCell *lc;
+	foreach (lc, subpaths)
+	{
+		Path *subpath = lfirst(lc);
+
+		/* Check if we have an append path under an append path (e.g., a partially compressed
+		 * chunk)*/
+		List *subsubpaths = get_subpaths_from_append_path(subpath, false);
```

Comment:
Hmm, would it make sense to call `generate_agg_pushdown_path` recursively here instead of special-casing this?

Reply:
I considered this. But I was not able to come up with an elegant solution that is shorter / better to read than special-casing this.